### PR TITLE
feat: get user's features, inclduing `expiresAt` field for each one

### PR DESCRIPTION
**Pull request by vardan-arm**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "@newrelic/native-metrics": "^6.0.0",
     "@standardnotes/auth": "3.2.0",
     "@standardnotes/domain-events": "1.4.1",
+    "@standardnotes/features": "^1.2.0",
```
Comment (karolsojko) [inline package.json:28]:
```suggestion
    "@standardnotes/features": "1.2.0",
```

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }))
+
+    return {
+      success: true,
+      userUuid,
+      features: features.flat() as Feature[],
+    }
+  }
+
+  private roleNameToSubscriptionNameMap = new Map<RoleName, SubscriptionName>([
```
Comment (karolsojko) [inline src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:61]: hmm we might extract that somewhere - I think @antsgar had the same thing

**Review by karolsojko — APPROVED** [review]
Good stuff 👍

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts (hunk 2 above)
Comment (vardan-arm) [reply src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:61]: I thought about that, @antsgar has the opposite mapper, probably we can put that logic in some shared place, I just didn't find such place in `auth`. Maybe it makes sense to store this in `snjs`, not sure though it's the best place.

File: package.json (hunk 1 above)
Comment (vardan-arm) [reply package.json:28]: Done

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        }
+      })
+
+      return featuresWithExpirationDate
+    }))
+
+    return {
+      success: true,
+      userUuid,
+      features: features.flat() as Feature[],
```
Comment (antsgar) [inline src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:57]: Perhaps we should make this a Set so that there aren't any duplicate features?

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    const userRoles = await user.roles
+    const userSubscriptions = await user.subscriptions
+
+    const features = await Promise.all(userRoles.map(async (role) => {
+      const subscriptionName = this.roleNameToSubscriptionNameMap.get(role.name as RoleName)
+      const expiresAt = (userSubscriptions.find(subscription => subscription.planName === subscriptionName) as UserSubscription).endsAt
+
+      const permissions = await role.permissions
+      const featuresWithExpirationDate = permissions.map(permission => {
+        const featureItem = Features.find(feature => feature.identifier === permission.name)
```
Comment (antsgar) [inline src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:43]: To make this a bit more performant, maybe we could first create a Set of the user's permissions, then iterate over features only once to check which features correspond to an existing user's permission (using filter for example)

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts (hunk 3 above)
Comment (vardan-arm) [reply src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:57]: Good point, it raises another question though: Let's say `Permission 1` belongs to `Role 1` and `Role 2`, expiration date of `Role 1` is September 2021 and for `Role 2` is March 2022. In such case I think we should set `expiresAt` for the feature that matches `Permission 1` to March 2022, is that correct?

Comment (karolsojko) [reply src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:57]: @vardan-arm correct.

TL;DR: longest subscription(role) that entitles you to a feature is the dominant one.

so the `Permission X` that has the longest expiration date should overwrite all `Permission X'` in the set that are it's duplicates with lesser expiration date.

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts (hunk 2 above)
Comment (karolsojko) [reply src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts:61]: hmm for now I would go with `src/Domain/Role/RoleToSubscriptionMap.ts` with the following methods `getSubscriptionNameForRoleName(...)`, `getRoleNameForSubscriptionName(...)`.

Comment (vardan-arm) [reply src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts]: Added, thanks)

File: src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts (hunk 4 above)
Comment (vardan-arm) [reply src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts]: Updated to be slightly performant along with removing duplications

**Comment by vardan-arm** [comment]
Closing this PR in favor of [this one](https://github.com/standardnotes/auth/pull/40), with verified commits.
